File: samples/java/Bingo/Server/Api/src/main/java/systems/zlink/samples/bingo/server/api/handlers/MatchBingoHandler.java
package systems.zlink.samples.bingo.server.api.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.channels.ZLinkRouteClient;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.ApiChannel)
public final class MatchBingoHandler
        implements ZLinkRequestHandler<Messages.MatchBingoApiReq, Messages.MatchBingoApiRes> {
    private final ZLinkRouteClient routes;
    private final ZLinkSpotManager spots;

    public MatchBingoHandler(ZLinkRouteClient routes, ZLinkSpotManager spots) {
        this.routes = routes;
        this.spots = spots;
    }

    @Override
    public CompletionStage<Messages.MatchBingoApiRes> handle(
            Messages.MatchBingoApiReq request, ZLinkMessageContext context) {
        // --8<-- [start:doc-bingo-api-match]
        String levelBucket = "1-10";
        return routes.requestToSpot(
                        "match:" + levelBucket,
                        BingoMessages.reserveBingoRoomReq(
                                request.getMode(), request.getActorId(), levelBucket))
                .instanceSpot(SampleNames.MatchmakerSpotType)
                .inMesh(SampleNames.MatchmakingMesh)
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.ReserveBingoRoomRes.class)
                .thenCompose(
                        allocated ->
                                spots.getOrCreate(allocated.getRoomId(), SampleNames.RoomSpotType)
                                        .inMesh(SampleNames.Mesh)
                                        .request(
                                                BingoMessages.bingoRoomCreateReq(
                                                        allocated.getSettings()))
                                        .timeout(SampleTimings.RequestTimeout)
                                        .submit()
                                        .thenApply(
                                                ignored ->
                                                        BingoMessages.matchBingoApiRes(
                                                                allocated.getRoomId())));
        // --8<-- [end:doc-bingo-api-match]
    }
}

File: samples/kotlin/Bingo/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/api/handlers/MatchBingoHandler.kt
package systems.zlink.samples.kotlin.bingo.server.api.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.channels.ZLinkRouteClient
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomCreateReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReserveBingoRoomReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReserveBingoRoomRes

@ZLinkHandlerGroup(SampleNames.ApiChannel)
class MatchBingoHandler(private val routes: ZLinkRouteClient, private val spots: ZLinkSpotManager) :
    ZLinkSuspendingRequestHandler<MatchBingoApiReq, MatchBingoApiRes> {
    override suspend fun handle(request: MatchBingoApiReq, context: ZLinkMessageContext) = run {
        // --8<-- [start:doc-bingo-api-match]
        val levelBucket = "1-10"
        val allocated =
            routes
                .requestToSpot(
                    "match:$levelBucket",
                    ReserveBingoRoomReq(request.actorId, request.mode, levelBucket),
                )
                .instanceSpot(SampleNames.MatchmakerSpotType)
                .inMesh(SampleNames.MatchmakingMesh)
                .timeout(SampleTimings.RequestTimeout)
                .submit(ReserveBingoRoomRes::class.java)
                .await()

        spots
            .getOrCreate(allocated.roomId, SampleNames.RoomSpotType)
            .inMesh(SampleNames.Mesh)
            .request(BingoRoomCreateReq(allocated.settings))
            .timeout(SampleTimings.RequestTimeout)
            .submit()
            .await()
        // --8<-- [end:doc-bingo-api-match]

        MatchBingoApiRes(allocated.roomId)
    }
}
